### ISSUE TITLE
reading std_out of non blocking commands

### ISSUE THREAD

**Issue by puterleat** [kennethreitz/delegator.py#24]
When I try:

```
c = delegator.run(['sleep', ' 10'], block=False)
c.out
```
The call blocks for 10 secs then produces "" as expected. However if I try:

```
c = delegator.run(['sleep', ' 10'], block=False)
c.std_out.read()
```

I get `UnsupportedOperation: not readable`

Is this expected? I had expected to be able to iterate over c.std_out to monitor progress.

**Comment by daylanKifky** [comment]
I'm having a similar problem, trying to get the output of a long running program that waits some time between writes.
something like:
```python
proc = delegator.run(['ping', '8.8.8.8'], block=False)

print proc.pid

print proc.out
```
would print the pid, then hangs outputting nothing.
has anyone figured this out?

**Comment by akshaykumar90** [comment]
I ran into the same problem and looked into the code.

@puterleat `delegator.py` uses the `pexpect`library for non-blocking commands and wires the `std_out` property to the `stdout` property of the underlying `PopenSpawn` object in `pexpect`. This does not work because that `stdout` **always** references `sys.stdout`.

Hence, the `UnsupportedOperation: not readable` exception, because you're trying to read from `sys.stdout` which only supports writing to.

@daylanKifky `proc.out` would not work either because it **always** blocks on the command to finish (even if you specify non-blocking). And since ping never terminates on successful connection, you will be kept waiting at `proc.out` forever. I think the behavior of `out` is correct, and is as-designed.

Anyway, I was able to read the stdout of a non-blocking command like below:

```python
c = delegator.run(cmd_string, block=False)
for line in c.subprocess:
    print(line.rstrip())
```

I think this is only a workaround. Ideally, reading via the `std_out` property should work for non blocking commands.
